File: learn/graph/dfs/dfs.go


type DFSGraph struct {
	vertices int
	adjList  [][]int
}

func NewDFSGraph(v int) *DFSGraph {
	return &DFSGraph{
		vertices: v,
		adjList:  make([][]int, v),
	}
}

func (g *DFSGraph) AddEdge(src, dest int) {
	g.adjList[src] = append(g.adjList[src], dest)
	g.adjList[dest] = append(g.adjList[dest], src)
}

func (g *DFSGraph) DFS(startVertex int) []int {
	visited := make([]bool, g.vertices)
	var result []int
	g.dfsUtil(startVertex, visited, &result)
	return result
}

func (g *DFSGraph) dfsUtil(vertex int, visited []bool, result *[]int) {
	visited[vertex] = true
	*result = append(*result, vertex)

	for _, neighbor := range g.adjList[vertex] {
		if !visited[neighbor] {
			g.dfsUtil(neighbor, visited, result)
		}
	}
}

File: learn/graph/dfs/dfs.rs
pub struct DFSGraph {
    vertices: usize,
    adj_list: Vec<Vec<usize>>,
}

impl DFSGraph {
    pub fn new(vertices: usize) -> Self {
        DFSGraph {
            vertices,
            adj_list: vec![Vec::new(); vertices],
        }
    }

    pub fn add_edge(&mut self, src: usize, dest: usize) {
        self.adj_list[src].push(dest);
        self.adj_list[dest].push(src); // undirected
    }

    pub fn dfs(&self, start_vertex: usize) -> Vec<usize> {
        let mut visited = vec![false; self.vertices];
        let mut result = Vec::new();
        self.dfs_util(start_vertex, &mut visited, &mut result);
        result
    }

    fn dfs_util(&self, vertex: usize, visited: &mut Vec<bool>, result: &mut Vec<usize>) {
        visited[vertex] = true;
        result.push(vertex);

        for &neighbor in &self.adj_list[vertex] {
            if !visited[neighbor] {
                self.dfs_util(neighbor, visited, result);
            }
        }
    }
}
